ProjFS.Mac/PrjFSKextLogDaemon/PrjFSKextLogDaemon.cpp

```diff
@@ -204,29 +265,222 @@ static bool TryFetchAndLogKextHealthData(io_connect_t connection)
+    healthDataWriter.Add("FindRootMisses", healthData.totalFindRootForVnodeMisses);
+    healthDataWriter.Add("RefreshRoot", healthData.totalRefreshRootForVnode);
+    healthDataWriter.Add("InvalidateRoot", healthData.totalInvalidateVnodeRoot);
+    WriteJsonToMessageListener(MessageType::Health, healthDataWriter);
+}
+
+static void CreatePipeToMessageListener()
+{
+    lock_guard<mutex> lock(s_messageListenerMutex);
+    
+    if (INVALID_SOCKET_FD != s_messageListenerSocket)
+    {
+        // Already connected
+        return;
+    }
+
+    int error = 0;
+
+    s_messageListenerSocket = socket(PF_UNIX, SOCK_STREAM, 0);
+    if (s_messageListenerSocket < 0)
+    {
+        error = errno;
+        os_log(s_daemonLogger, "CreatePipeToMessageListener: Failed to create socket, error: %d", error);
+        s_messageListenerSocket = INVALID_SOCKET_FD;
+        return;
+    }
+    
+    struct sockaddr_un socket_address;
+    memset(&socket_address, 0, sizeof(struct sockaddr_un));
+    
+    socket_address.sun_family = AF_UNIX;
+    size_t pathLength = MessageListenerSocketPath.length();
+    if (pathLength + 1 >= sizeof(socket_address.sun_path))
+    {
+        os_log(
+            s_daemonLogger,
+            "CreatePipeToMessageListener: Failed to copy socket path (%{public}s), insufficient buffer. Buffer size: %zu",
+            MessageListenerSocketPath.c_str(),
+            sizeof(socket_address.sun_path));
+        goto ClosePipeAndCleanup;
+    }
+    
+    strlcpy(socket_address.sun_path, MessageListenerSocketPath.c_str(), sizeof(socket_address.sun_path));
+    
+    if(0 == connect(s_messageListenerSocket, (struct sockaddr *) &socket_address, sizeof(struct sockaddr_un)))
+    {
+        os_log(
+            s_daemonLogger,
+            "CreatePipeToMessageListener: Connected to message listener at %{public}s",
+            MessageListenerSocketPath.c_str());
+        return;
+    }
+    
+    error = errno;
+    os_log(
+        s_daemonLogger,
+        "CreatePipeToMessageListener: Failed to connect socket at %{public}s, error: %d",
+        MessageListenerSocketPath.c_str(),
+        error);
+    
+ClosePipeAndCleanup:
+    ClosePipeToMessageListener_Locked();
+}
+
+static void ClosePipeToMessageListener_Locked()
+{
+    if (INVALID_SOCKET_FD != s_messageListenerSocket)
+    {
+        close(s_messageListenerSocket);
+        s_messageListenerSocket = INVALID_SOCKET_FD;
+    }
+}
+
+static string MessageTypeToString(MessageType messageType)
+{
+    switch (messageType) {
+        case MessageType::Info:
+            return "info";
+            
+        case MessageType::Error:
+            return "error";
+            
+        case MessageType::Health:
+            return "health";
+    }
+    
+    return "invalid";
+}
+
+static void WriteJsonToMessageListener(MessageType messageType, const JsonWriter& jsonMessage)
+{
+    lock_guard<mutex> lock(s_messageListenerMutex);
```

Comment:
Putting the `write()` inside a held lock could cause everything to get backed up - if reads stall on the other side of the socket for some reason (I/O…) that will fill up the socket's internal buffer, which will block the `write()` here, which means the mutex sits in the held state.

I don't think there's a great catch-all solution to this, just something to be aware of.

Reply:
Yep that is definitely a risk, however, as best I could tell a backup here should not cause a backup in the kext itself (although messages might get dropped).

If we find that we do encounter problems with the reader getting stuck one option is moving all writes to a new background thread (which will help ensure we don't block reading messages from the kext and writing them to the macOS log), but I will leave that for a follow up PR if we end up needing it.